### PR TITLE
[Balance] list historical balances

### PR DESCRIPTION
**Pull request by xinyu-li-cb**
- [PSDK-360] Support gasless sends (#130)
- [Balance] list historical balances

### What changed? Why?
Add function `historical_balances` for Address, which fetches historical balances of an asset for an address

SDK test:
[1] pry(main)> Coinbase.configure_from_json("/Users/xinyu.li@coinbase.com/ext/coinbase-sdk-nodejs/coinbase_cloud_api_key.json")
=> "Successfully configured Coinbase SDK"
[2] pry(main)> address = Coinbase::Address.new("base-sepolia","0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789")
=> Coinbase::Address{id: '0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789', network_id: 'base_sepolia'}
[3] pry(main)> address.historical_balances('usdc').first
=> Coinbase::HistoricalBalance{amount: '36', block_height: '11576200', block_hash: '0x2a059cfec6397b9d57dea43d7336a2236c18e33113eba690db657bbc9f377c3d', asset: 'Coinbase::Asset{network_id: 'base_sepolia', asset_id: 'usdc', decimals: '6', address_id: '0x036CbD53842c5426634e7929541eC2318f3dCF7e'}'}

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>1</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>2</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>1</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>2</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: lib/coinbase/address.rb [hunk 1]
```diff
@@ -62,6 +62,27 @@ def balance(asset_id)
       Coinbase::Balance.from_model_and_asset_id(response, asset_id).amount
     end
 
+    # Enumerates the historical balances for a given asset belonging of address.
+    # The result is an enumerator that lazily fetches from the server, and can be iterated over,
+    # converted to an array, etc...
+    # @return [Enumerable<Coinbase::HistoricalBalance>] Enumerator that returns historical_balance
+    def hiatorical_balances(asset_id)
```
Comment (alex-stone) [inline lib/coinbase/address.rb:69]:
```suggestion
    def historical_balances(asset_id)
```

File: lib/coinbase/address.rb [hunk 2]
```diff
@@ -62,6 +62,27 @@ def balance(asset_id)
...
+      Coinbase::Pagination.enumerate(
+        ->(page) { list_page(asset_id, page) }
+      ) do |historical_balance|
+        Coinbase::HistoricalBalance.from_model(historical_balance)
+      end
+    end
+
+    def list_page(asset_id, page)
```
Comment (alex-stone) [inline lib/coinbase/address.rb:77]: This should be beneath the `private` directive

File: lib/coinbase/address.rb (hunk 1 above)
Comment (xinyu-li-cb) [reply lib/coinbase/address.rb:69]: let me correct it. thanks for catching!

File: lib/coinbase/historical_balance.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      @asset = asset
+    end
+
+    attr_reader :amount, :block_height, :block_hash, :asset
+
+    # Returns a string representation of the HistoricalBalance.
+    # @return [String] a string representation of the HistoricalBalance
+    def to_s
+      "Coinbase::Balance{amount: '#{amount.to_i}', block_height: '#{block_height.to_i}', " \
+       "block_hash: '#{block_hash}', asset: '#{asset}'}"
```
Comment (alex-stone) [inline lib/coinbase/historical_balance.rb:39]:
```suggestion
    Coinbase.pretty_print_object(
        self.class,
        amount: amount.to_i
        block_height: block_height.to_i
        block_hash: block_hash,
        asset: asset
    )
```

File: lib/coinbase/address.rb (hunk 2 above)
Comment (xinyu-li-cb) [reply lib/coinbase/address.rb:77]: thanks for catching! let me move it.

File: lib/coinbase/address.rb (hunk 1 above)
Comment (alex-stone) [inline lib/coinbase/address.rb:69]: Can we add a spec for this to our `shared_examples/address_balances`?

**Comment by alex-stone** [comment]
Can you update the Changelog.md and update the PR section `What changed? Why?` Thanks!

File: lib/coinbase/address.rb (hunk 1 above)
Comment (xinyu-li-cb) [reply lib/coinbase/address.rb]: added into the balance session, thanks a lot!

**Review by alex-stone — APPROVED** [review]

**Pull request merged**